modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultStaticConstantExtractionPhase.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.painless.phase;
+
+import org.elasticsearch.painless.ir.ClassNode;
+import org.elasticsearch.painless.ir.ConstantNode;
+import org.elasticsearch.painless.ir.FieldNode;
+import org.elasticsearch.painless.symbol.IRDecorations.IRDConstant;
+import org.elasticsearch.painless.symbol.IRDecorations.IRDConstantFieldName;
+import org.elasticsearch.painless.symbol.IRDecorations.IRDExpressionType;
+import org.elasticsearch.painless.symbol.IRDecorations.IRDFieldType;
+import org.elasticsearch.painless.symbol.IRDecorations.IRDModifiers;
+import org.elasticsearch.painless.symbol.IRDecorations.IRDName;
+import org.elasticsearch.painless.symbol.ScriptScope;
+
+import java.lang.reflect.Modifier;
+
+/**
```

Comment:
Here is the secret magic of this PR: if you define a `ConstantNode` you used to have to attach a constant that'd fit in the "constant pool". See the `instanceof` tree below for what fits. This finds what doesn't fit and causes us to inject the values when we build the script. This is a trick we can do because painless is always embedded in a running process so we can just build what we need at compile time and stick it on the script objects we build. `javac` has to write class files and it wouldn't have a place to "put" that state. We're lucky to be embedded, making this fairly easy for us.

Reply:
Not for this PR, but would it make sense to combine this into the constant folding phase rather than have a separate phase for it?

Reply:
I wanted to run this "after" all of the constants have resolved, just in case some constant doesn't fit in the constant pool but we end up removing it during the constant folding phase.